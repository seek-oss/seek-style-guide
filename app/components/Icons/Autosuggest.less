@import (reference) "~seek-style-guide/theme";

@border-color: @sk-mid-grey-light;
@border-radius: 4px;

.container {
  position: relative;
}

.input {
  .touchableText();
  width: @grid-gutter-width * 14;
  height: @grid-row-height * 5;
  padding: @grid-row-height @grid-gutter-width;
  border: 1px solid @border-color;
  border-radius: @border-radius;

  &:focus {
    outline: none;
  }

  .containerOpen & {
    border-bottom-left-radius: 0;
    border-bottom-right-radius: 0;
  }
}

.suggestionsContainer {
  position: absolute;
  top: @grid-row-height * 5 - 1px;
  width: @grid-gutter-width * 14;
  margin: 0;
  padding: 0;
  list-style-type: none;
  border: 1px solid @border-color;
  background-color: @sk-white;
  border-bottom-left-radius: @border-radius;
  border-bottom-right-radius: @border-radius;
  z-index: 2;
  max-height: @grid-row-height * 29;
  overflow-y: auto;
}

.suggestion {
  cursor: pointer;
  padding: @grid-row-height @grid-gutter-width;
}

.suggestionFocused {
  background-color: @sk-mid-grey-light;
}

.iconSvg {
  fill: @sk-charcoal;
  width: (@grid-row-height * 2);
  height: (@grid-row-height * 2);
}

.suggestionText {
  .standardText();
  padding-left: @grid-gutter-width;
}
