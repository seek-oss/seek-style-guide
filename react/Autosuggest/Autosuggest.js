import styles from './Autosuggest.less';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import ReactAutosuggest from '../private/react-autosuggest';
import IsolatedScroll from 'react-isolated-scroll';
import omit from 'lodash/omit';
import TextField from '../TextField/TextField';
import smoothScroll from '../private/smoothScroll';
import smallDeviceOnly from '../private/smallDeviceOnly';

/* eslint-disable react/no-deprecated */
export default class Autosuggest extends Component {
  static displayName = 'Autosuggest';

  static propTypes = {
    id: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    type: PropTypes.string,
    inputProps: PropTypes.object,
    label: PropTypes.string,
    labelProps: PropTypes.object,
    className: PropTypes.string,
    autosuggestProps: PropTypes.object.isRequired,
    showMobileBackdrop: PropTypes.bool,
    /* eslint-disable consistent-return */
    suggestionsContainerClassName: (props, _, componentName) => {
      const { suggestionsContainerClassName, autosuggestProps } = props;
      const { theme } = autosuggestProps || {};
      const { suggestionsContainer } = theme || {};
      const propType = typeof suggestionsContainerClassName;

      if (propType !== 'undefined' && propType !== 'string') {
        return new Error(
          `Invalid prop \`suggestionsContainerClassName\` of type \`${propType}\` supplied to \`${componentName}\`, expected \`string\`.`
        );
      }

      if (suggestionsContainer && suggestionsContainerClassName) {
        return new Error(
          '`suggestionsContainerClassName` will be overridden by the `suggestionsContainer` class in autosuggestProps `theme`. Please remove it.'
        );
      }
    }
    /* eslint-enable consistent-return */
  };

  static defaultProps = {
    type: 'text',
    className: '',
    label: '',
    labelProps: {},
    inputProps: {},
    showMobileBackdrop: false
  };

  state = {
    areSuggestionsShown: false
  };

  componentWillUpdate(nextProps, nextState) {
    if (
      !this.state.areSuggestionsShown &&
      nextState.areSuggestionsShown &&
      smallDeviceOnly()
    ) {
      smoothScroll(this.textField);
    }
  }

  storeInputReference = autosuggest => {
    if (autosuggest !== null) {
      this.input = autosuggest.input;
    }
  };

  storeTextFieldReference = textField => {
    if (textField !== null) {
      this.textField = textField.container;
    }
  };

  renderSuggestionsContainer = ({ containerProps, children }) => {
    const { ref, ...rest } = containerProps;
    const areSuggestionsShown = children !== null;

    if (this.state.areSuggestionsShown !== areSuggestionsShown) {
      this.setState({ areSuggestionsShown });
    }

    const callRef = isolatedScroll => {
      if (isolatedScroll !== null) {
        ref(isolatedScroll.component);
      }
    };

    return <IsolatedScroll {...rest} ref={callRef} children={children} />;
  };

  renderInputComponent = inputProps => {
    const { label, labelProps } = this.props;

    const textFieldProps = {
      ref: this.storeTextFieldReference,
      inputProps,
      ...omit(this.props, ['inputProps', 'labelProps', 'autosuggestProps'])
    };

    if (label) {
      textFieldProps.labelProps = {
        ...labelProps,
        className: classnames({
          [styles.isLabelCoveredWithBackdrop]: this.props.showMobileBackdrop,
          [labelProps.className]: labelProps.className
        })
      };
    }

    return <TextField {...textFieldProps} />;
  };

  render() {
    const {
      value,
      onChange,
      onFocus,
      onBlur,
      type,
      inputProps,
      label,
      autosuggestProps,
      suggestionsContainerClassName,
      showMobileBackdrop
    } = this.props;
    const { theme = {} } = autosuggestProps;

    const allAutosuggestProps = {
      renderSuggestionsContainer: this.renderSuggestionsContainer,
      renderInputComponent: this.renderInputComponent,
      ...autosuggestProps,
      theme: {
        ...styles,
        suggestionsContainer: classnames({
          [styles.suggestionsContainer]: true,
          [styles.suggestionsContainer_withLabel]: label,
          [suggestionsContainerClassName]: suggestionsContainerClassName
        }),
        ...theme
      }
    };

    return (
      <div>
        <ReactAutosuggest
          inputProps={{ value, onChange, onFocus, onBlur, type, ...inputProps }}
          ref={this.storeInputReference}
          {...allAutosuggestProps}
        />
        <div
          className={classnames({
            [styles.autosuggestBackdrop]: true,
            [styles.backdrop_isMobile]: showMobileBackdrop
          })}
        />
      </div>
    );
  }
}
