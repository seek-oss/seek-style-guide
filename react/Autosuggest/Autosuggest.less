@import (reference) "~seek-style-guide/theme";

@suggestionContainerRowsOffset: 9;
@suggestionFieldHeightInRows: 8;

@sectionTitleHeight: @row-height * 6;
@containerHeight: @touchableTextHeight + @sectionTitleHeight;

.container {
  position: relative;
  display: inline-block;
  width: 100%;
}

.containerOpen {
  z-index: @z-index-page-overlay;
}

.suggestionsContainer {
  display: none;
  position: absolute;
  background-color: @sk-white;
  box-shadow: 0 9px 30px rgba(0, 0, 0, 0.4);
  padding: @row-height 0;
  border-radius: 2px;
  min-width: 100%;
  max-width: 100%;
  max-height: 385px;
  overflow-y: auto;
  top: @row-height * @suggestionContainerRowsOffset;

  .smallDeviceOnly({max-height: ~"calc(100vh - @{containerHeight})" ;});
}

.suggestionsContainerOpen {
  display: block;
}

.suggestionsContainer_withLabel {
  @rows-below-standard-text: 1;
  @labelRowOffset: @standard-type-row-span + @rows-below-standard-text;
  top: @row-height * (@suggestionContainerRowsOffset + @labelRowOffset);
}

.sectionTitle {
  .rawText(1);
  line-height: @sectionTitleHeight;
  height: @sectionTitleHeight;
  margin: @row-height (@gutter-width * 0.75) 0;
  color: @sk-form-accent;
  font-weight: @sk-bold;

  .sectionContainer:first-child & {
    margin-top: 0;
  }
}

.suggestion {
  .rawText();
  line-height: (@row-height * @suggestionFieldHeightInRows);
  height: (@row-height * @suggestionFieldHeightInRows);
  padding: 0 (@gutter-width * 0.75);
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  cursor: pointer;
}

.suggestionHighlighted {
  background-color: @sk-mid-gray-light;
}

.autosuggestBackdrop {
  visibility: hidden;
  .transparentBackdrop();
}

.containerOpen + .autosuggestBackdrop {
  visibility: visible;

  &.backdrop_isMobile {
    .smallDeviceOnly({.backdrop() ;});
  }
}

.containerOpen .isLabelCoveredWithBackdrop {
  display: block;
  pointer-events: none;
}
