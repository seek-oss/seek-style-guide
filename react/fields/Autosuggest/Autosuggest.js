import textfieldStyle from '../TextField/TextField.less';
import autosuggestStyles from './Autosuggest.less';

import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import ReactAutosuggest from 'react-autosuggest';

import ErrorIcon from '../../icons/ErrorIcon/ErrorIcon';
import ClearField from '../TextField/ClearField/ClearField';

function combineClassNames(props = {}, ...classNames) {
  const { className, ...restProps } = props;

  return {
    className: classnames.apply(null, [...classNames, className]), // eslint-disable-line no-useless-call
    ...restProps
  };
}

export default class Autosuggest extends Component {

  static displayName = 'Autosuggest';

  static propTypes = {
    /* eslint-disable consistent-return */
    id: (props, propName, componentName) => {
      const { id, label } = props;

      if (typeof id !== 'string') {
        return new Error(`Invalid prop \`id\` of type \`${typeof id}\` supplied to \`${componentName}\`, expected \`string\`.`);
      }

      if (label && !id) {
        return new Error(`When ${componentName} has a \`label\`, it should also have an \`id\`.`);
      }
    },
    /* eslint-enable consistent-return */
    className: PropTypes.string,
    invalid: PropTypes.bool,
    label: PropTypes.string,
    /* eslint-disable consistent-return */
    labelProps: (props, propName, componentName) => {
      const { id, label, labelProps } = props;
      const { htmlFor: labelFor } = labelProps || {};

      if (typeof labelProps !== 'undefined' && typeof labelProps !== 'object') {
        return new Error(`Invalid prop \`labelProps\` of type \`${typeof labelProps}\` supplied to \`${componentName}\`, expected \`object\`.`);
      }

      if (!label && labelProps) {
        return new Error(`Specifying \`labelProps\` is redundant when \`label\` is not specified in ${componentName}.`);
      }

      if (labelFor && id) {
        return new Error(`\`labelProps.htmlFor\` will be overridden by \`id\` in ${componentName}. Please remove it.`);
      }
    },
    inputProps: (props, propName, componentName) => {
      const { id, inputProps } = props;
      const { id: inputId } = inputProps || {};

      if (typeof inputProps !== 'undefined' && typeof inputProps !== 'object') {
        return new Error(`Invalid prop \`inputProps\` of type \`${typeof inputProps}\` supplied to \`${componentName}\`, expected \`object\`.`);
      }

      if (inputId && id) {
        return new Error(`\`inputProps.id\` will be overridden by \`id\` in ${componentName}. Please remove it.`);
      }
    },
    /* eslint-disable consistent-return */
    autosuggestProps: PropTypes.object.isRequired,
    help: PropTypes.string,
    helpProps: PropTypes.object,
    message: PropTypes.string,
    messageProps: PropTypes.object,
    onClear: PropTypes.func
  };

  static defaultProps = {
    id: '',
    className: '',
    invalid: false,
    label: '',
    help: '',
    message: ''
  };

  constructor() {
    super();

    this.storeInputReference = this.storeInputReference.bind(this);
    this.renderLabel = this.renderLabel.bind(this);
    this.renderInput = this.renderInput.bind(this);
    this.renderHelp = this.renderHelp.bind(this);
    this.renderMessage = this.renderMessage.bind(this);
    this.renderIcon = this.renderIcon.bind(this);
    this.handleMouseDownOnClear = this.handleMouseDownOnClear.bind(this);
  }

  storeInputReference(autosuggest) {
    if (autosuggest !== null) {
      this.input = autosuggest.input;
    }
  }

  handleMouseDownOnClear(event) {
    const { onClear } = this.props;

    if (typeof onClear === 'function') {
      onClear(event);
      this.input.focus();
      event.preventDefault(); // https://developer.mozilla.org/en/docs/Web/API/HTMLElement/focus#Notes
    }
  }

  renderLabel() {
    const { label } = this.props;

    if (!label) {
      return;
    }

    const { labelProps, id } = this.props;
    const allLabelProps = {
      ...combineClassNames(labelProps, textfieldStyle.label),
      ...(id ? { htmlFor: id } : {})
    };

    return (
      <label {...allLabelProps}>
        {label}
      </label>
    );
  }

  renderInput() {
    const { inputProps, id, onClear, autosuggestProps } = this.props;
    const { suggestions, onSuggestionsFetchRequested, theme, onSuggestionsClearRequested,
            onSuggestionSelected, getSuggestionValue, renderSuggestion, shouldRenderSuggestions } = autosuggestProps;
    const allInputProps = {
      ...(id ? { id } : {}),
      ...inputProps
    };
    const autosuggestTheme = {
      input: classnames({
        [textfieldStyle.input]: true,
        [textfieldStyle.input_isClearable]: onClear
      }),
      ...(theme || autosuggestStyles)
    };

    return (
      <ReactAutosuggest
        inputProps={allInputProps}
        suggestions={suggestions}
        onSuggestionsFetchRequested={onSuggestionsFetchRequested}
        onSuggestionsClearRequested={onSuggestionsClearRequested}
        onSuggestionSelected={onSuggestionSelected}
        getSuggestionValue={getSuggestionValue}
        renderSuggestion={renderSuggestion}
        shouldRenderSuggestions={shouldRenderSuggestions}
        theme={autosuggestTheme}
        ref={this.storeInputReference}
      />
    );
  }

  renderClear() {
    return (
      <span
        className={textfieldStyle.clearField}
        onMouseDown={this.handleMouseDownOnClear}
        onTouchStart={this.handleMouseDownOnClear}>
        <ClearField />
      </span>
    );
  }

  renderHelp() {
    const { message, help } = this.props;

    if (message || !help) {
      return;
    }

    const { helpProps } = this.props;
    const allHelpProps = combineClassNames(helpProps, textfieldStyle.help);

    return (
      <p {...allHelpProps}>
        {help}
      </p>
    );
  }

  renderMessage() {
    const { message } = this.props;

    if (!message) {
      return;
    }

    const { messageProps } = this.props;
    const allMessageProps = combineClassNames(messageProps, textfieldStyle.message);

    return (
      <p {...allMessageProps}>
        {this.renderIcon()}
        {message}
      </p>
    );
  }

  renderIcon() {
    const { invalid } = this.props;

    if (!invalid) {
      return;
    }

    return (
      <ErrorIcon
        filled={true}
        className={textfieldStyle.messageIcon}
        svgClassName={textfieldStyle.messageIconSvg}
      />
    );
  }

  render() {
    const { className, invalid, onClear, inputProps = {} } = this.props;
    const hasValue = (inputProps.value && inputProps.value.length > 0);
    const canClear = hasValue && (typeof onClear === 'function');
    const classNames = classnames({
      [textfieldStyle.root]: true,
      [textfieldStyle.invalid]: invalid,
      [textfieldStyle.canClear]: canClear,
      [className]: className
    });

    return (
      <div className={classNames}>
        {this.renderLabel()}
        {this.renderInput()}
        {this.renderClear()}
        {this.renderHelp()}
        {this.renderMessage()}
      </div>
    );
  }

}
