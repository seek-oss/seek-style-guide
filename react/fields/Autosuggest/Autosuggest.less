@import (reference) "~seek-style-guide/theme";

.container {
  position: relative;
  display: inline-block;
  width: 100%;
}

@suggestionContainerRowsOffset: 6;
.suggestionsContainer {
  display: none;
  z-index: 1;
  position: absolute;
  background-color: @sk-white;
  box-shadow: 0 9px 30px rgba(0, 0, 0, 0.4);
  padding: @grid-row-height 0;
  border-radius: 2px;
  min-width: 100%;
  max-width: 100%;
  max-height: 385px;
  overflow-y: auto;
  top: @grid-row-height * @suggestionContainerRowsOffset;

  .containerOpen & {
    display: block;
  }
}

.suggestionsContainer_withLabel {
  @labelRowOffset: 4;
  top: @grid-row-height * (@suggestionContainerRowsOffset + @labelRowOffset);

  @media only screen and (min-width: (@responsive-breakpoint)) {
    @labelRowOffset: 3;
    top: @grid-row-height * (@suggestionContainerRowsOffset + @labelRowOffset);
  }
}

.sectionTitle {
  .rawText(1);
  line-height: (@grid-row-height * 4);
  height: (@grid-row-height * 4);
  margin: @grid-row-height (@grid-gutter-width * 0.75) 0;
  color: darken(@sk-highlight, 10%);
  font-weight: @sk-bold;

  .sectionContainer:first-child & {
    margin-top: 0;
  }
}

.suggestion {
  .rawText();
  line-height: (@grid-row-height * 4);
  height: (@grid-row-height * 4);
  padding: 0 (@grid-gutter-width * 0.75);
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  cursor: pointer;
}

.suggestionFocused {
  background-color: @sk-mid-gray-light;
}
