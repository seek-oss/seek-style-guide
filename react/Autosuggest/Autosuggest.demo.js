import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Autosuggest from './Autosuggest';
import styles from '../TextField/TextField.less';
import classnames from 'classnames';
import fieldMessageOptions from '../FieldMessage/FieldMessage.demoFragment';
import fieldLabelOptions from '../FieldLabel/FieldLabel.demoFragment';

class AutosuggestContainer extends Component {
  static propTypes = {
    component: PropTypes.func.isRequired,
    componentProps: PropTypes.object.isRequired
  };

  constructor() {
    super();

    this.state = {
      inputValue: ''
    };
  }

  handleChange = (_, { newValue }) => {
    this.setState({
      inputValue: newValue
    });
  };

  handleFocus = event => {
    console.log('Focused, event: ', event);
  };

  handleClear = () => {
    this.setState({
      inputValue: ''
    });
  };

  render() {
    const { component: DemoComponent, componentProps } = this.props;
    const { inputValue } = this.state;

    return (
      <div style={{ width: '300px' }}>
        <DemoComponent
          {...componentProps}
          type="search"
          value={inputValue}
          onChange={this.handleChange}
          onFocus={this.handleFocus}
          onClear={this.handleClear}
        />
      </div>
    );
  }
}

export default {
  route: '/autosuggest',
  title: 'Autosuggest',
  category: 'Form',
  component: Autosuggest,
  container: AutosuggestContainer,
  initialProps: {
    id: 'jobTitles',
    label: 'Job Titles',
    type: 'search',
    value: '...',
    onChange: () => {},
    autosuggestProps: {
      suggestions: [
        'Developer',
        'Product manager',
        'Iteration manager',
        'Designer'
      ],
      onSuggestionsFetchRequested: () => {},
      onSuggestionsClearRequested: () => {},
      renderSuggestion: suggestion => <div>{suggestion}</div>,
      getSuggestionValue: suggestion => suggestion
    },
    message: false,
    onClear: () => {}
  },
  options: [
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Focus',
          transformProps: ({ className, ...props }) => ({
            ...props,
            className: classnames(className, styles.rootFocus)
          })
        },
        {
          label: 'Grouped',
          transformProps: ({ autosuggestProps, ...props }) => ({
            ...props,
            autosuggestProps: {
              ...autosuggestProps,
              multiSection: true,
              suggestions: [
                {
                  title: 'RECENT TITLES',
                  suggestions: [
                    'Developer',
                    'Product manager',
                    'Iteration manager',
                    'Designer'
                  ]
                }
              ],
              renderSectionTitle: section => <div>{section.title}</div>,
              getSectionSuggestions: section => section.suggestions
            }
          })
        },
        {
          label: 'Show mobile backdrop',
          transformProps: ({ ...props }) => ({
            ...props,
            showMobileBackdrop: true
          })
        }
      ]
    },
    ...fieldMessageOptions,
    ...fieldLabelOptions
  ]
};
