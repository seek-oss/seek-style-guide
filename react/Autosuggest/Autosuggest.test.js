import { mount, render } from 'enzyme';
import React from 'react';
import Autosuggest from './Autosuggest';

const getAutosuggestProps = (suggestions = []) => ({
  autosuggestProps: {
    suggestions,
    onSuggestionsFetchRequested: () => {},
    onSuggestionsClearRequested: () => {},
    getSuggestionValue: s => s,
    renderSuggestion: s => <span className="suggestion">{s}</span>,
    onSuggestionSelected: () => {},
    alwaysRenderSuggestions: true
  },
  value: '',
  onChange: () => {}
});

describe('Autosuggest', () => {
  let spy;

  beforeEach(() => {
    spy = jest.spyOn(global.console, 'error').mockImplementation(() => {}); // Swallow console errors
  });

  afterEach(() => {
    spy.mockRestore();
  });

  it('should render with simple props', () => {
    const wrapper = render(
      <Autosuggest {...getAutosuggestProps()} id="testAutosuggest" />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass through the type', () => {
    const wrapper = mount(
      <Autosuggest
        {...getAutosuggestProps()}
        type="search"
        id="testAutosuggest"
      />
    );
    const inputType = wrapper.find('input').prop('type');
    expect(inputType).toEqual('search');
  });

  it('should pass through the value', () => {
    const wrapper = mount(
      <Autosuggest
        {...getAutosuggestProps()}
        value="foo"
        id="testAutosuggest"
      />
    );
    const inputValue = wrapper.find('input').prop('value');
    expect(inputValue).toEqual('foo');
  });

  it('should render with suggestions', () => {
    const wrapper = render(
      <Autosuggest
        {...getAutosuggestProps(['test', 'test 2'])}
        id="testAutosuggest"
      />
    );
    const suggestions = wrapper
      .find('.Autosuggest__suggestionsContainer')
      .find('li');
    expect(wrapper).toMatchSnapshot();
    expect(suggestions.length).toEqual(2);
  });

  it('should render with label', () => {
    const wrapper = render(
      <Autosuggest
        {...getAutosuggestProps()}
        id="testAutosuggest"
        label="Foo"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with mobile backdrop', () => {
    const props = getAutosuggestProps();
    props.labelProps = {
      className: 'LABEL_TEST_CLASS'
    };

    const wrapper = render(
      <Autosuggest
        {...props}
        id="testAutosuggest"
        label="Foo"
        showMobileBackdrop
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should error if `suggestionsContainerClassName` is not a string', () => {
    const expectedError = expect.stringMatching(
      /Invalid prop `suggestionsContainerClassName` of type `boolean` supplied to `Autosuggest`, expected `string/
    );
    render(
      <Autosuggest
        {...getAutosuggestProps()}
        id="testAutosuggest"
        suggestionsContainerClassName={true}
      />
    );

    expect(spy).toBeCalledWith(expectedError);
  });

  it('should error if `suggestionsContainerClassName` and theme with class `suggestionsContainer` are both specified', () => {
    const props = getAutosuggestProps();
    props.autosuggestProps.theme = {
      suggestionsContainer: 'TEST'
    };
    const expectedError = expect.stringMatching(
      /`suggestionsContainerClassName` will be overridden by the `suggestionsContainer` class in autosuggestProps `theme`. Please remove it./
    );

    render(
      <Autosuggest
        {...props}
        id="testAutosuggest"
        suggestionsContainerClassName="TEST 2"
      />
    );
    expect(spy).toBeCalledWith(expectedError);
  });

  it('should focus field when suggestion is clicked', () => {
    const wrapper = mount(
      <Autosuggest
        {...getAutosuggestProps(['test', 'test 2'])}
        id="testAutosuggest"
      />
    );
    const firstSuggestion = wrapper.find('li').first();
    const input = wrapper.find('input').html();
    firstSuggestion.simulate('click');
    expect(global.document.activeElement.outerHTML).toEqual(input);
  });

  it('should invoke the focus handler', () => {
    const callback = jest.fn();
    const wrapper = mount(
      <Autosuggest
        {...getAutosuggestProps()}
        onFocus={callback}
        id="testAutosuggest"
      />
    );
    wrapper.find('input').simulate('focus');
    expect(callback.mock.calls.length).toEqual(1);
  });

  it('should invoke the blur handler', () => {
    const callback = jest.fn();
    const wrapper = mount(
      <Autosuggest
        {...getAutosuggestProps()}
        onBlur={callback}
        id="testAutosuggest"
      />
    );
    wrapper.find('input').simulate('blur');
    expect(callback.mock.calls.length).toEqual(1);
  });

  it('should invoke the change handler', () => {
    const callback = jest.fn();
    const wrapper = mount(
      <Autosuggest
        {...getAutosuggestProps()}
        onChange={callback}
        id="testAutosuggest"
      />
    );
    wrapper.find('input').simulate('change', { target: { value: 'foo' } });
    expect(callback.mock.calls.length).toEqual(1);
    expect(callback.mock.calls[0][0].target.value).toEqual('foo');
  });

  describe('inputProps', () => {
    it('should pass through value', () => {
      const wrapper = mount(
        <Autosuggest
          {...getAutosuggestProps()}
          inputProps={{ value: 'input props value' }}
          id="testAutosuggest"
        />
      );
      const inputValue = wrapper.find('input').prop('value');
      expect(inputValue).toEqual('input props value');
    });

    it('should invoke the focus event', () => {
      const onFocus = jest.fn();
      const wrapper = mount(
        <Autosuggest
          {...getAutosuggestProps()}
          inputProps={{ onFocus }}
          id="testAutosuggest"
        />
      );
      wrapper.find('input').simulate('focus');
      expect(onFocus.mock.calls.length).toEqual(1);
    });

    it('should invoke the blur event', () => {
      const onBlur = jest.fn();
      const wrapper = mount(
        <Autosuggest
          {...getAutosuggestProps()}
          inputProps={{ onBlur }}
          id="testAutosuggest"
        />
      );
      wrapper.find('input').simulate('blur');
      expect(onBlur.mock.calls.length).toEqual(1);
    });

    it('should invoke the change event', () => {
      const onChange = jest.fn();
      const wrapper = mount(
        <Autosuggest
          {...getAutosuggestProps()}
          inputProps={{ onChange }}
          id="testAutosuggest"
        />
      );
      wrapper.find('input').simulate('change', { target: { value: 'foo' } });
      expect(onChange.mock.calls.length).toEqual(1);
      expect(onChange.mock.calls[0][0].target.value).toEqual('foo');
    });
  });
});
