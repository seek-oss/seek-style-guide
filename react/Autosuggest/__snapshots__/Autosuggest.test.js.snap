// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autosuggest should render with label 1`] = `
<div>
  <div
    class="Autosuggest__container"
  >
    <div
      class="TextField__root"
    >
      <label
        class=""
        for="testAutosuggest"
      >
        <span
          class="Text__root Text__standard Text__baseline"
        >
          <span
            class=""
          >
            <strong
              class="Strong__root"
            >
              Foo
            </strong>
             
          </span>
        </span>
      </label>
      <input
        aria-autocomplete="list"
        aria-describedby="testAutosuggest-message"
        aria-expanded="false"
        aria-owns="react-autowhatever-1"
        autocomplete="off"
        class="TextField__input"
        id="testAutosuggest"
        role="combobox"
        type="text"
        value=""
      />
      <span
        class="TextField__clearField"
      >
        <div
          class="ClearField__root"
        >
          <span
            class="Icon__root"
          >
            mock svg
          </span>
        </div>
      </span>
      <div
        class="FieldMessage__root"
      />
    </div>
    <div
      class="Autosuggest__suggestionsContainer Autosuggest__suggestionsContainer_withLabel"
      id="react-autowhatever-1"
    />
  </div>
  <div
    class="Autosuggest__autosuggestBackdrop"
  />
</div>
`;

exports[`Autosuggest should render with mobile backdrop 1`] = `
<div>
  <div
    class="Autosuggest__container"
  >
    <div
      class="TextField__root"
    >
      <label
        class="Autosuggest__isLabelCoveredWithBackdrop LABEL_TEST_CLASS"
        for="testAutosuggest"
      >
        <span
          class="Text__root Text__standard Text__baseline"
        >
          <span
            class=""
          >
            <strong
              class="Strong__root"
            >
              Foo
            </strong>
             
          </span>
        </span>
      </label>
      <input
        aria-autocomplete="list"
        aria-describedby="testAutosuggest-message"
        aria-expanded="false"
        aria-owns="react-autowhatever-1"
        autocomplete="off"
        class="TextField__input"
        id="testAutosuggest"
        role="combobox"
        type="text"
        value=""
      />
      <span
        class="TextField__clearField"
      >
        <div
          class="ClearField__root"
        >
          <span
            class="Icon__root"
          >
            mock svg
          </span>
        </div>
      </span>
      <div
        class="FieldMessage__root"
      />
    </div>
    <div
      class="Autosuggest__suggestionsContainer Autosuggest__suggestionsContainer_withLabel"
      id="react-autowhatever-1"
    />
  </div>
  <div
    class="Autosuggest__autosuggestBackdrop Autosuggest__backdrop_isMobile"
  />
</div>
`;

exports[`Autosuggest should render with simple props 1`] = `
<div>
  <div
    class="Autosuggest__container"
  >
    <div
      class="TextField__root"
    >
      <input
        aria-autocomplete="list"
        aria-describedby="testAutosuggest-message"
        aria-expanded="false"
        aria-owns="react-autowhatever-1"
        autocomplete="off"
        class="TextField__input"
        id="testAutosuggest"
        role="combobox"
        type="text"
        value=""
      />
      <span
        class="TextField__clearField"
      >
        <div
          class="ClearField__root"
        >
          <span
            class="Icon__root"
          >
            mock svg
          </span>
        </div>
      </span>
      <div
        class="FieldMessage__root"
      />
    </div>
    <div
      class="Autosuggest__suggestionsContainer"
      id="react-autowhatever-1"
    />
  </div>
  <div
    class="Autosuggest__autosuggestBackdrop"
  />
</div>
`;

exports[`Autosuggest should render with suggestions 1`] = `
<div>
  <div
    class="Autosuggest__container Autosuggest__containerOpen"
  >
    <div
      class="TextField__root"
    >
      <input
        aria-autocomplete="list"
        aria-describedby="testAutosuggest-message"
        aria-expanded="true"
        aria-owns="react-autowhatever-1"
        autocomplete="off"
        class="TextField__input"
        id="testAutosuggest"
        role="combobox"
        type="text"
        value=""
      />
      <span
        class="TextField__clearField"
      >
        <div
          class="ClearField__root"
        >
          <span
            class="Icon__root"
          >
            mock svg
          </span>
        </div>
      </span>
      <div
        class="FieldMessage__root"
      />
    </div>
    <div
      class="Autosuggest__suggestionsContainer Autosuggest__suggestionsContainerOpen"
      id="react-autowhatever-1"
    >
      <ul
        role="listbox"
      >
        <li
          class="Autosuggest__suggestion"
          data-suggestion-index="0"
          id="react-autowhatever-1--item-0"
          role="option"
        >
          <span
            class="suggestion"
          >
            test
          </span>
        </li>
        <li
          class="Autosuggest__suggestion"
          data-suggestion-index="1"
          id="react-autowhatever-1--item-1"
          role="option"
        >
          <span
            class="suggestion"
          >
            test 2
          </span>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="Autosuggest__autosuggestBackdrop"
  />
</div>
`;
